chain/messagepool/messagepool.go

```diff
@@ -494,6 +509,12 @@ func (mp *MessagePool) getStateBalance(addr address.Address, ts *types.TipSet) (
 }
 
 func (mp *MessagePool) PushWithNonce(ctx context.Context, addr address.Address, cb func(address.Address, uint64) (*types.SignedMessage, error)) (*types.SignedMessage, error) {
+	// serialize push access to reduce lock contention
+	mp.addSema <- struct{}{}
+	defer func() {
+		<-mp.addSema
+	}()
```

Comment:
PushWith nonce calls a callback that signs a message, so it would be great if we didn't hold the mp lock through the signign operation but this might be hard as signing operation can fail...

Reply:
Yeah, that's a problem indeed. Let me see if we can do it in a reasonable fashion.

Reply:
I think we can temporarily release the locks before signing and then reacquire, but then we'll have to verify that the state hasn't changed in the meantime.
If that happens, we can retry the whole thing.

Reply:
or instead of retrying fail with EAGAIN or something equivalent.

Reply:
I relinquished the locks while signing, and reacquired with state check later on.
If the state check fails, `PushWithNonce` returns `ErrTryAgain` which is caught by the api call to retry.